routers/web/repo/repo.go

```diff
@@ -576,23 +577,61 @@ func SearchRepo(ctx *context.Context) {
 		return
 	}
 
-	results := make([]*api.Repository, len(repos))
+	// collect the latest commit of each repo
+	repoIDsWithCommitStatuses := make([]int64, 0, len(repos))
+	commitsWithCommitStatuses := make([]string, 0, len(repos))
+	for _, repo := range repos {
+		branches, branchCount, err := repo_service.GetBranches(ctx, repo, 0, 0)
+		if err != nil {
+			log.Error("GetBranches: %v", err)
+			continue
+		}
+		if branchCount == 0 {
+			continue
+		}
+
+		for _, branch := range branches {
+			// find the default branch
+			if repo.DefaultBranch == branch.Name {
+				commit, err := branch.GetCommit()
+				if err != nil {
+					log.Error("GetCommit: %v", err)
+					break
+				}
+				repoIDsWithCommitStatuses = append(repoIDsWithCommitStatuses, repo.ID)
+				commitsWithCommitStatuses = append(commitsWithCommitStatuses, commit.ID.String())
+				break
+			}
+		}
+	}
+
+	// call the database O(1) times to get the commit statuses for all repos
+	repoToItsLatestCommitStatuses, err := git_model.GetLatestCommitStatusForPairs(ctx, repoIDsWithCommitStatuses, commitsWithCommitStatuses, db.ListOptions{})
+	if err != nil {
+		log.Error("GetLatestCommitStatusForPairs: %v", err)
+		return
+	}
+
+	results := make([]*api.WebSearchRepository, len(repos))
 	for i, repo := range repos {
-		results[i] = &api.Repository{
-			ID:       repo.ID,
-			FullName: repo.FullName(),
-			Fork:     repo.IsFork,
-			Private:  repo.IsPrivate,
-			Template: repo.IsTemplate,
-			Mirror:   repo.IsMirror,
-			Stars:    repo.NumStars,
-			HTMLURL:  repo.HTMLURL(),
-			Link:     repo.Link(),
-			Internal: !repo.IsPrivate && repo.Owner.Visibility == api.VisibleTypePrivate,
+		results[i] = &api.WebSearchRepository{
+			Repository: &api.Repository{
+				ID:       repo.ID,
+				FullName: repo.FullName(),
+				Fork:     repo.IsFork,
+				Private:  repo.IsPrivate,
+				Template: repo.IsTemplate,
+				Mirror:   repo.IsMirror,
+				Stars:    repo.NumStars,
+				HTMLURL:  repo.HTMLURL(),
+				Link:     repo.Link(),
+				Internal: !repo.IsPrivate && repo.Owner.Visibility == api.VisibleTypePrivate,
+			},
+			LatestCommitStatusState: git_model.CalcCommitStatus(repoToItsLatestCommitStatuses[repo.ID]).State,
```

Comment:
If we already calculate the whole commit status, we might as well return the whole status instead of only its state…

Reply:
Wouldn't this problem also be solved by moving your new `modules` structs into `services`?

Reply:
Good thinking, done, thanks